### ISSUE TITLE
zipstream not file-like enough

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2996]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#2996|
|Type|enhancement|
|Created|2008-01-13 06:27:10Z|
|Branch|https://github.com/twisted/twisted/tree/filezipfile-2996|

I was writing some code that uses `FilePath` and I discovered that `twisted.python.zipstream` advertises a file-like interface but is actually missing:

  * `__iter__()`
  * `next()`
  * `closed`
  * `isatty`
  * `readline()`
  * `readlines()`
  * `xreadlines()`

It might also be nice to have `seek()`, since there is a `tell()`.

<details><summary>Searchable metadata</summary>

```
trac-id__2996 2996
type__enhancement enhancement
reporter__glyph glyph
priority__highest highest
milestone__ 
branch__branches_filezipfile_2996 branches/filezipfile-2996
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1200205630000000 1200205630000000
changetime__1202372634000000 1202372634000000
version__None None
owner__ 
cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

(In [22335]) Branching to 'filezipfile-2996'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph removed owner|
|-|-|

This turned into a bit of a free-for-all cleanup, but the test coverage was so crappy before that I didn't feel comfortable touching anything.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

Taking back for some 2.3-2.4 compatibility fixes...

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

OK, the appropriate tests are just set to skip in versions prior to 2.5; the zipfile module just does something different in its constructor, so there's no way to get to the offending codepath.  It's an unfortunate compatibility, but it's not code we maintain...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

Let's go!

zipstream.py:
 * it needs a copyright statement.
 * the module docstring should be reformated.
 * the from future import can be removed.
 * ```_FileEntry``` could be a new style class.
 * its init method should document its parameters. Its instance variables should be documented too.
 * the finished attribute of ```_FileEntry``` should be a boolean.
 * the readline method seems really inefficient. I don't know how to improve it though :).
 * unzipIter could use filepath (not to be done here).
 * in unzipIter, you should write ```remaining -= 1```.
 * in countZipFileChunks, you should write ```totalchunks += etc```.
 * countFileChunks could have a docstring. It could use the divmod function too.
 * countZipFileEntries should have a deprecation warning.
 * the docstring of countZipFileEntries should use `C{str}`, not just str
 * unzipIterChunky should use filepath, use ```remaining -= 1```, and should not have assert

test_zipstream.py:
 * it should have a docstring and a copyright notice
 * most docstrings could have some epydoc love (use of `L{}` and `C{}`)
 * I don't see why junk and junkmd5 are defined at module level
 * test_unzipIter uses range(0, x), it should use range(x). It should write some paces around ```numfiles-i-1```
 * test_closed should assertTrue and assertFalse
 * test_isatty should use assertFalse
 * test_unzip should put some spaces around ```['']*numfiles```

Overall, this is mostly nitpicking, this is a huge improvement. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Replying to [therve](#note_6):
> Let's go!
>
> zipstream.py:
>  * its init method should document its parameters. Its instance variables should be documented too.

The convention thus far (to avoid requiring silly levels of redundancy) has been to document the ivars and leave them out of the `__init__` signature if they're simple assignments, so I've done that.

>  * the finished attribute of ```_FileEntry``` should be a boolean.

This is left over from the previous implementation, so I'm leaving it as it is.

>  * the readline method seems really inefficient. I don't know how to improve it though :).

Not something I'm really worried about right now.

>  * unzipIter could use filepath (not to be done here).
>  * unzipIterChunky should use filepath, use ```remaining -= 1```, and should not have assert

I think that the right thing to do here would be to eventually make a `copyTo` that works with `zippath` and just delete unzipIter and friends.  (asserts deleted, since they are eternally valueless)

> test_zipstream.py:
>  * test_closed should assertTrue and assertFalse
>  * test_isatty should use assertFalse

I intentionally avoid these for better error reporting when something goes wrong.  I want the values to be literally equal to True and False.  For example, if it starts returning '7', or '""', something is weird, even though that will pass an `assertTrue` and `assertFalse`

> Overall, this is mostly nitpicking, this is a huge improvement. Thanks!

How's it looking now? :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Replying to [glyph](#note_7):
>
> I intentionally avoid these for better error reporting when something goes wrong.  I want the values to be literally equal to True and False.  For example, if it starts returning '7', or '""', something is weird, even though that will pass an `assertTrue` and `assertFalse`

If this is valid for this code, isn't it valid for all code?  If so, then shouldn't ```assertTrue``` and ```assertFalse``` be removed, or at least never used in Twisted itself?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_8):
> If this is valid for this code, isn't it valid for all code?  If so, then shouldn't ```assertTrue``` and ```assertFalse``` be removed, or at least never used in Twisted itself?

This is more a matter of personal style than a matter of policy.  There are cases where it seems appropriate; for example, when the assertion being performed isn't the system under test.  In this code, `.closed` is the SUT and should therefore be closely examined.  But, for example, in `test_strcred` there are assertions like `self.assertTrue(ICredentialsChecker.providedBy(checker))`.  `assertTrue` would be out of place if that were a test for Zope Interface, but in that context it seems fine to me.

(Of course, an even better assertion would be an `assertProvides` which computed the list of interfaces which were provided and formatted a message describing the object, its type, any `directlyProvides` interfaces that it had specified, and gave you all of that information in the assertion message.  But I digress.  A lot.)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @glyph|
|-|-|

 * _FileEntry still doesn't document its fp, finished, and closes attributes
 * in unzipIterChunky, there are still some ```remaining = remaining-1```
 * in the same, there is also a one liner ```if not os.path.exists(f): os.makedirs(f)```

You can merge once corrected :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

(In [22463]) re [#2996](https://github.com/twisted/twisted/issues/2996) last few review tweaks.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set status to `closed`|
|-|-|

(In [22464]) Make the file-like objects returned by zipstream substantially more file-like.

This change adds the `readline` method, the iterator protocol, `readlines`, `xreadlines`, and the `closed` attribute to the objects in zipstream.

It also vastly improves the quality of implementation, adds a bunch more tests, and makes a ton of coding-standard compliance changes.

Author: glyph

Reviewer: therve

Fixes [#2996](https://github.com/twisted/twisted/issues/2996)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
